aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1315,81 +1315,349 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
+        }
+
+        auto t_counts = at::ones_like(t_unique);
+        int64_t len_unique;
+        {
+          auto* ptr_counts = t_counts.data_ptr<int64_t>();
+          auto* first = t_unique.data_ptr<int64_t>();
+          auto* last = first + len;
+          auto* curr = first;
+          auto* counts = ptr_counts;
+          while (++first != last) {
+            if (*curr != *first) {
+              ++counts;
+              if (++curr != first) {
+                *curr = *first;
+              }
+            }
+            else {
+              ++(*counts);
+            }
+          }
+          len_unique = counts - ptr_counts + 1;
+        }
+
+        return std::make_tuple(t_idx, t_unique, t_counts, len_unique);
+      };
+
+      Tensor dim_sort_indices, dim_indices_unique, dim_indices_counts;
+      int64_t n_unique_dim_indices;
+      std::tie(dim_sort_indices, dim_indices_unique, dim_indices_counts, n_unique_dim_indices)
+        = unique_with_counts(
+            indices[dim], nnz,
+            // if self is coalesced and dim == 0,
+            // indices[dim] is already sorted
+            /*is_sorted=*/self.is_coalesced() && dim == 0
+          );
+
+      Tensor sel_sort_indices, sel_indices_unique, sel_indices_counts;
+      int64_t n_unique_sel_indices;
+      std::tie(sel_sort_indices, sel_indices_unique, sel_indices_counts, n_unique_sel_indices)
+        = unique_with_counts(nneg_index, index_len);
+
+      // index intersection is aka `set(indices[dim]) - set(index)`.
+      // If `dim_indices_unique[i] == sel_indices_unique[j]`, then
+      // nnz += dim_indices_counts[i] * sel_indices_unique[j]
+      const auto compute_index_intersections_and_nnz = [](
+          const Tensor& t1, const Tensor& c1, int64_t l1,
+          const Tensor& t2, const Tensor& c2, int64_t l2
+      ) -> std::tuple<Tensor, Tensor, int64_t, int64_t> {
+        const auto lmin = std::min(l1, l2);
+        auto t1_idx = at::empty({lmin}, t1.options());
+        auto t2_idx = at::empty({lmin}, t2.options());
+        int64_t nnz = 0;
+
+        auto* ptr_t1_idx = t1_idx.data_ptr<int64_t>();
+        auto* ptr_t2_idx = t2_idx.data_ptr<int64_t>();
+
+        auto* ptr_t1 = t1.data_ptr<int64_t>();
+        auto* ptr_c1 = c1.data_ptr<int64_t>();
+        auto* ptr_t2 = t2.data_ptr<int64_t>();
+        auto* ptr_c2 = c2.data_ptr<int64_t>();
+
+        // we assume search in t2
+        auto* first = ptr_t2;
+        auto* last = ptr_t2 + l2;
+
+        for (const auto i : c10::irange(l1)) {
+          const auto idx = ptr_t1[i];
+          const auto idx_pos = std::lower_bound(first, last, idx);
+          if (idx_pos != last && *idx_pos == idx) {
+            const auto j = idx_pos - first;
+            const auto count1 = ptr_c1[i];
+            const auto count2 = ptr_c2[j];
+            *ptr_t1_idx++ = i;
+            *ptr_t2_idx++ = j;
+            nnz += count1 * count2;
+          }
+        }
```

Comment:
This look can also be done in parallel with `nnz = (count1 * count2).sum().item()`.